framework/src/main/java/org/checkerframework/common/returnsreceiver/ReturnsReceiverVisitor.java

```diff
@@ -0,0 +1,44 @@
+package org.checkerframework.common.returnsreceiver;
+
+import com.sun.source.tree.*;
+import com.sun.source.util.TreePath;
+import javax.lang.model.element.AnnotationMirror;
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.common.basetype.BaseTypeVisitor;
+import org.checkerframework.javacutil.AnnotationUtils;
+import org.checkerframework.javacutil.TreeUtils;
+
+/** The visitor for the Returns Receiver Checker. */
+public class ReturnsReceiverVisitor extends BaseTypeVisitor<ReturnsReceiverAnnotatedTypeFactory> {
+
+    /**
+     * Create a new {@code ReturnsReceiverVisitor}.
+     *
+     * @param checker the type-checker associated with this visitor
+     */
+    public ReturnsReceiverVisitor(BaseTypeChecker checker) {
+        super(checker);
+    }
+
+    @Override
+    public Void visitAnnotation(AnnotationTree node, Void p) {
+        AnnotationMirror annot = TreeUtils.annotationFromAnnotationTree(node);
+        // Warn if a @This annotation is in an illegal location.
+        if (AnnotationUtils.areSame(annot, getTypeFactory().THIS_ANNOTATION)) {
+            TreePath parentPath = getCurrentPath().getParentPath();
+            Tree parent = parentPath.getLeaf();
+            Tree grandparent = parentPath.getParentPath().getLeaf();
+            boolean isReturnAnnot =
+                    grandparent instanceof MethodTree
+                            && (parent.equals(((MethodTree) grandparent).getReturnType())
+                                    || parent instanceof ModifiersTree);
+            boolean isCastAnnot =
+                    grandparent instanceof TypeCastTree
+                            && parent.equals(((TypeCastTree) grandparent).getType());
+            if (!(isReturnAnnot || isCastAnnot)) {
```

Comment:
I find `!isReturn && !isCastAnnot` easier to look at visually and easier to understand... but you might prefer this way.

Reply:
I personally like it this way 🙂 I think we will leave it

Reply:
Sure, no objection.